text/0000-string-copy_cpointer-copy_cstring-returns-iso.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: Modify String.copy\_cpointer() and String.copy\_cstring() to return iso^
+- Start Date: 2025-09-10
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+The two current constructors `String.copy_cpointer()` and `String.copy_cstring()` both take a `Pointer[U8] box` and return a `String ref^`. Given that these two constructors make copies, returning a `String iso^` would be preferable and safe.
+
+Note: In the interests of clarity of writing, I will use `String.copy_cstring()` throughout the rest of this RFC. All of the commentary below should be read to apply to both `String.copy_cstring()`, and `String.copy_cpointer()`.
+
+# Motivation
+
+In many cases, as a library author I want to provide my end-users a mutable and sendable copy of a String received via C-FFI. In other words, a `String iso^`. With the way that the String library is currently written, there are two ways to do this:
+
+## Create a String ref^ using String.from\_cstring() or from\_cpointer() and clone().
+
+```pony
+  var str: String iso = String.from_cstring(ptr).clone()
+```
+
+The main issue with this is that it directly violates the documentation which states: "This must be done only with C-FFI functions that return pony\_alloc'd character arrays".
+
+## Create a String ref^ using String.copy\_cstring() and clone()
```

Comment:
again, here you are doing a clone and that is wasteful. You create a string solely to clone it to get a new string. eeewww. unclean! unclean!

Reply:
Agreed - this is the motivation for the change…  To avoid this second copy.